leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java

```diff
@@ -159,18 +158,44 @@ protected Object handleInvocation(Object proxy, Method method, Object[] args) th
 
     private RuntimeException handleDelegateThrewException(
             LeadershipToken leadershipToken, InvocationTargetException exception) throws Exception {
-        if (exception.getCause() instanceof ServiceNotAvailableException
-                || exception.getCause() instanceof NotCurrentLeaderException) {
-            leadershipStateManager.handleNotLeading(leadershipToken, exception.getCause());
-        }
+        Throwable cause = exception.getCause();
+
+        maybeHandleLeadershipChangeException(leadershipToken, cause);
+
         // Prevent blocked lock requests from receiving a non-retryable 500 on interrupts
         // in case of a leader election.
-        if (exception.getCause() instanceof InterruptedException
-                && !leadershipCoordinator.isStillCurrentToken(leadershipToken)) {
+        if (cause instanceof InterruptedException && !leadershipCoordinator.isStillCurrentToken(leadershipToken)) {
             throw leadershipCoordinator.notCurrentLeaderException(
                     "received an interrupt due to leader election.", exception.getCause());
         }
-        Throwables.propagateIfPossible(exception.getCause(), Exception.class);
-        throw new RuntimeException(exception.getCause());
+        Throwables.propagateIfPossible(cause, Exception.class);
+        throw new RuntimeException(cause);
+    }
+
+    /**
+     * Check and see if the provided throwable thew a {@link ServiceNotAvailableException} and if so, check and see
+     * if it is a {@link MaybeNotCurrentLeaderException}. If it is, check and see if we are still leading, as we may
+     * still be, but contending with ourselves. Otherwise, assume we are not, and invalidate our leadership state.
+     */
+    private void maybeHandleLeadershipChangeException(LeadershipToken leadershipToken, Throwable throwable)
+            throws ExecutionException, InterruptedException {
+        if (throwable instanceof ServiceNotAvailableException) {
+            boolean notLeading = true;
+
+            if (throwable instanceof MaybeNotCurrentLeaderException) {
+                notLeading = isNotLeading(
+                        leadershipCoordinator.isStillLeading(leadershipToken).get());
+            }
+
+            if (notLeading) {
+                leadershipStateManager.handleNotLeading(leadershipToken, throwable);
+            }
+        }
+    }
+
+    private static boolean isNotLeading(StillLeadingStatus leading) {
```

Comment:
This feels weird... but it doesn't really make sense to put this elsewhere. The enum is purposefully designed in this way to force users to deal with the `NO_QUORUM` state, but for us we don't care about it. Or such that -- we see NO_QUORUM and NOT_LEADING as functionally equivalent.